### PR TITLE
Disable parse options

### LINKED ISSUE #19: Disable Parsing Options
Since Marine is a per-packet parser (and doesn't support conversations and fragmentation), we can disable some options to speed up parsing and save some memory. This is a list of TCP options (available through Edit ->Preferences -> Protocols):
image

Need to check how tshark supports disabling options and apply the logic here. We should be able to disable as many options as possible (without affecting the parsing results).

Some options which can definitely be disabled:
- Analyze TCP sequence numbers
- Relative sequence numbers
- Track number of bytes in flight
- Reassemble out-of-order segments
- Validate the TCP checksum if possible

There are also options for IPv4, IPv6, etc. Need to scan them all and disable as many as possible.

### PR DESCRIPTION
**Pull request by royreznik**
resolve #19

### CODE REVIEW

File: marine.c
Comment (yehonatanz) [inline marine.c]: `fprintf(stderr, ...)`

File: marine.c
Comment (yehonatanz) [inline marine.c]: `fprintf(stderr, ...)`

File: marine.c
Comment (yehonatanz) [inline marine.c]: That's a weird choice of return value.
I'd rather you return the number of failures, that will be more useful and will have the added bonus of being zero when the function succeeds completely.

File: marine.c
Comment (yehonatanz) [inline marine.c]: Remove this unnecessary declaration

File: marine.c
Comment (yehonatanz) [inline marine.c]: `"preference used to be exist"` -> `"preference is obselete"`

File: marine.h
Comment (yehonatanz) [inline marine.h]: Rename to `marine_pref_t`, it took me a while to figure out that's not a wireshark type defined in `prefs.h`

**Review by yehonatanz — CHANGES_REQUESTED** [review]
Looks great.
I do want to wait for a steady benchmark before merging it.

File: marine.h
Comment (royreznik) [reply marine.h]: Its because you are stupid

File: marine_benchmark.c
Comment (yehonatanz) [inline marine_benchmark.c]: `ARRAY_SIZE(three_fields)` etc.

**Review by yehonatanz — CHANGES_REQUESTED** [review]

**Review by yehonatanz — APPROVED** [review]
Approved.
Merge is blocked by a steady benchmark

**Comment by tomlegkov** [comment]
> Approved.
> Merge is blocked by a steady benchmark

What would a steady benchmark require that we don't currently have? How do we measure it? @yehonatanz

File: marine.c
Comment (tomlegkov) [inline marine.c]: Taken

File: marine.c
Comment (tomlegkov) [inline marine.c]: Clearly the wrong way to increment a variable!

File: marine.c
Comment (tomlegkov) [inline marine.c]: Internal error

File: marine.c
Comment (tomlegkov) [inline marine.c]: I think we should also expose a way to change preferences through the API (maybe as arguments to `marine_init`?). Otherwise, every option change will require re-compiling Marine (we should still disable by default though), and maybe some client will want to enable something that we chose to disable.

File: marine.c
Comment (tomlegkov) [inline marine.c]: doesn't exist

**Review by tomlegkov — CHANGES_REQUESTED** [review]
Can you include output of the benchmark before and after disabling options?

File: marine.c
Comment (royreznik) [reply marine.c]: You can use this function whenever you want, it doesn't matter if marine already initialized
it's like using the `-o` flag in tshark

File: marine.c
Comment (royreznik) [reply marine.c]: what do you mean?

**Comment by royreznik** [comment]
> Can you include output of the benchmark before and after disabling options?

We are waiting for aws server so i can run the benchmark on clean state machine

File: marine.c
Comment (tomlegkov) [reply marine.c]: `failed++;`

File: marine.c
Comment (tomlegkov) [reply marine.c]: Doesn't `-o` in tshark only affect the initialization process? If it's really possible to use this after initialization too, I think a test showing it is required.

Also, in terms of exposing this as an API - the function isn't `WS_DLL_PUBLIC`, and I'm not sure how you can pass a `marine_pref_t*` object from ctypes (you can, but it's kinda useless since it's basically a string).

File: marine.c
Comment (royreznik) [reply marine.c]: Why does it matter?

File: marine.c
Comment (royreznik) [reply marine.c]: I will test it in the python client and update if it works as i thought it work

**Review by yehonatanz — APPROVED** [review]

File: marine.c
Comment (tomlegkov) [inline marine.c:267]: I think the API's output can be improved.
1. Printing to stderr isn't very useful. In Python it might be easy to catch it (which you don't currently do), but in other languages it might be much more difficult. Consider doing it like in `marine_add_filter` and writing the output to a buffer provided by the user.
2. The return value isn't very c-ish. I don't see how the amount of failed prefs is useful - I think a boolean is good enough with an error message that is easy to parse (array of strings? csv? what ever you like)

File: marine.c
Comment (tomlegkov) [inline marine.c:866]: If this fails, the init process should stop and return an error code.

**Review by tomlegkov — CHANGES_REQUESTED** [review]

File: marine.c
Comment (tomlegkov) [inline marine.c:230]: I think you can disable more reassembly options in IPv4, IPv6 and HTTP (and if you think of more common protocols for this, awesome).

**Comment by tomlegkov** [comment]
Did this PR change our existing benchmark speed? 

If not maybe we should adjust the benchmarks to show that this actually helps?

File: marine.c
Comment (shaharyarn) [inline marine.c:273]: Should be `using` instead of `use`.
Also the error message can sometimes not make sense if the error is for example `preference doesn't exist` and yet the message says it is using the default value for it.

File: marine.c
Comment (shaharyarn) [inline marine.c:278]: You need to free `err` (remember to use `g_free`).

**Review by shaharyarn — CHANGES_REQUESTED** [review]
Just two minor errors.

**Comment by royreznik** [comment]
will we reopen soon

**Pull request closed without merging**
